### PR TITLE
Minor type safety improvements

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Mostly, this consists of removing unnecessary use of `T.unsafe`. I'll add comments for the exceptions.

### CODE REVIEW

File: Library/Homebrew/cask/cask_loader.rb
Comment (dduugg) [inline Library/Homebrew/cask/cask_loader.rb:60]: Slight update to https://github.com/Homebrew/brew/pull/18019/files now that URI::Generic [supports](https://github.com/sorbet/sorbet/pull/8093) `to_str`

File: Library/Homebrew/version.rb
Comment (dduugg) [inline Library/Homebrew/version.rb:737]: per https://ruby-doc.org/3.3.4/Object.html#method-i-respond_to-3F `false` is the correct default

**Review by carlocab — APPROVED** [review]

**Review by p-linnane — APPROVED** [review]

File: Library/Homebrew/version.rb
Comment (Bo98) [reply Library/Homebrew/version.rb:737]: Maybe add `respond_to?` to `Style/OptionalBooleanParameter`'s `AllowedMethods`?

Comment (Bo98) [reply Library/Homebrew/version.rb:737]: Though I'm not sure if the use of this is intentional over `respond_to_missing?` which that cop already allows.

Comment (MikeMcQuaid) [reply Library/Homebrew/version.rb:737]: agreed 👍🏻

**Review by MikeMcQuaid — APPROVED** [review]
Looks good thanks @dduugg! Will hold off merging until after next tag 🔜 today.

**Comment by MikeMcQuaid** [comment]
Tagged, good to go when you are @dduugg!

File: Library/Homebrew/version.rb
Comment (dduugg) [reply Library/Homebrew/version.rb]: Done

**Pull request merged**
